Date picker layout exceptions

darrenaustin opened the pull request:
## Description

This PR addresses several layout issues with the Material Date Picker when displaying on smaller sized devices or with an increased text scale factor. It removes several hard coded sizes that were causing the layouts to overflow. This change will lead to some awkward layouts with scrolling the month grid vertically, but it should no longer throw overflow exceptions. We will address smaller layouts more fully in a future update.

### Example of layout fixes

Before/After on a Pixel with Large Text scale selected:

![Before After Pixel Large Text](https://user-images.githubusercontent.com/19588/56922569-56217400-6a7d-11e9-9ae5-2b9b6ea86db8.png)

Before/After on a Pixel with a Chinese locale:

![Before After Pixel Chinese](https://user-images.githubusercontent.com/19588/56922217-83215700-6a7c-11e9-8e3b-48b1aaf7d67a.png)

Before/After on a small display (320x521) with Large Text scale selected:

![Before After Small Display Large Text](https://user-images.githubusercontent.com/19588/56922851-055e4b00-6a7e-11e9-90c2-7cc15e40e069.png)

## Related Issues

Fixes: #21383, #20171, #19744, #17745

## Tests

I added the following tests:

- A group of device configuration tests that will try to bring up the date picker in various sized screens, orientations and text scale factors.

- A group of device configuration tests to the flutter_localizations tests to bring up the date picker in various locales that were having issues with layout.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
Fixed several layout issues with the material date picker. Mostly just removed hard coded sizes to allow the grid view to scroll instead of overflowing.

darrenaustin committed:
Merge remote-tracking branch 'upstream/master' into date_picker_layout

darrenaustin committed:
Added tests for smaller layouts and localizations.

darrenaustin committed:
Simplified the tests and removed some over-engineered configuration stuff.

HansMuller commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -682,7 +676,6 @@ class _MonthPickerState extends State<MonthPicker> with SingleTickerProviderStat
   @override
   Widget build(BuildContext context) {
     return SizedBox(
-      width: _kMonthPickerPortraitWidth,
```
Looking at this through the github review microsope ... it seems odd to configure the height of the MonthPicker with _kMaxDayPickerHeight. If that's intentional, maybe a comment that explains why.

darrenaustin replied:
Yeah, this is a bit weird. The MonthPicker just wraps the DayPicker with horizontal navigation to switch months. I added a comment that hopefully makes this easier to follow.

HansMuller commented on packages/flutter/test/material/date_picker_test.dart:
```diff
@@ -774,3 +777,70 @@ void _tests() {
+
+void _screenConfigTests() {
+
+  const Size kPixel1Portrait = Size(1070, 1770);
```
A comment about these sizes, like what they're based on, why we chose them, would be helpful.

darrenaustin replied:
Just pushed a fix that renamed them and commented them with hopefully better descriptions.

HansMuller commented on packages/flutter/test/material/date_picker_test.dart:
```diff
@@ -774,3 +777,70 @@ void _tests() {
 }
+
+void _screenConfigTests() {
+
```
Please insert a comment (probably several) like this here:
```
// Regression test for https://github.com/flutter/flutter/issues/21383
```

Or, if it makes more sense, in the specific `testWidgets` tests.

darrenaustin replied:
Done.

HansMuller commented on packages/flutter/test/material/date_picker_test.dart:
```diff
@@ -774,3 +777,70 @@ void _tests() {
+  }
+
+  testWidgets('display on Pixel1 - portrait', (WidgetTester tester) async {
+    await _showPicker(tester, kPixel1Portrait);
```
These tests are verifying that we don't assert. Just to make that clear it would be good to append:
```
 expect(tester.takeException(), isNull);
```

darrenaustin replied:
Nice suggestion. I have added this to all the tests.

HansMuller commented on packages/flutter_localizations/test/date_picker_test.dart:
```diff
@@ -223,6 +223,59 @@ void main() {
     await tester.tap(find.text('ANNULER'));
   });
+
+  group('screen configurations', () {
```
What these tests have in common seems to be that we're varying locale?

darrenaustin replied:
I renamed this group to "locale fonts don't overflow layout". Couldn't think of a better description. Let me know if you have one :).

HansMuller commented on packages/flutter_localizations/test/date_picker_test.dart:
```diff
@@ -223,6 +223,59 @@ void main() {
+
+    const Size kPixel1Portrait = Size(1070, 1770);
+    const Size kPixel1Landscape = Size(1770, 1070);
+    Future<void> _showPicker(WidgetTester tester, Locale locale, Size size) async {
```
add a blank line here?

darrenaustin replied:
Done.

HansMuller commented on packages/flutter_localizations/test/date_picker_test.dart:
```diff
@@ -223,6 +223,59 @@ void main() {
+
+  group('screen configurations', () {
+
+    const Size kPixel1Portrait = Size(1070, 1770);
```
As before, maybe there's something to say about the sizes

darrenaustin replied:
Done.

HansMuller reviewed (commented):
This looks good, just some minor feedback. 

It would be helpful if the issue's description included a few before/after screenshots that illustrated the layout changes.

darrenaustin committed:
Merge remote-tracking branch 'upstream/master' into date_picker_layout

darrenaustin commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -493,6 +486,7 @@ class DayPicker extends StatelessWidget {
             child: GridView.custom(
               gridDelegate: _kDayPickerGridDelegate,
               childrenDelegate: SliverChildListDelegate(labels, addRepaintBoundaries: false),
+              padding: EdgeInsets.zero,
```
I don't think so, as this GridView is embedded in a PageView.builder which puts the dialog header above it. If this padding isn't set to zero, by default there is an unneeded gap at the top of the grid, which will make it more likely that it will need to scroll on smaller screens.

darrenaustin committed:
Merge remote-tracking branch 'upstream/master' into date_picker_layout

darrenaustin committed:
Updates from PR review:
- Renamed the sizes used for testing and documented them with comments.
- Added more comments around the tests to better describe them and
  reference the bugs they test for.

darrenaustin committed:
Removed some final hard code sizes and opted instead to have the landscape picker always try to take up 2/3 of the space and give the other 3rd to the header.

HansMuller reviewed (approved):
LGTM
